src/System.Security.Cryptography.X509Certificates/tests/CertificateCreation/SubjectAltNameBuilderTests.cs

```diff
@@ -0,0 +1,266 @@
+
+            X509Extension extension = builder.Build();
+            Assert.Equal(SubjectAltNameOid, extension.Oid.Value);
+
+            Assert.Equal(
+                "301C821A786E2D2D7767763731616F3039652E6578616D706C652E6F7267",
+                extension.RawData.ByteArrayToHex());
+        }
+
+        [Fact]
+        public static void SingleValue_EmailAddress_Ascii()
+        {
+            SubjectAlternativeNameBuilder builder = new SubjectAlternativeNameBuilder();
+            builder.AddEmailAddress("user@example.org");
+
+            X509Extension extension = builder.Build();
+            Assert.Equal(SubjectAltNameOid, extension.Oid.Value);
+
+            Assert.Equal("3012811075736572406578616D706C652E6F7267", extension.RawData.ByteArrayToHex());
+        }
+
+        [Fact]
+        public static void SingleValue_EmailAddress_Unicode()
+        {
+            // There's not a good example of what an IDNA-converted email address
+            // looks like, so this isn't easy to verify.  For now let it be restricted to IA5.
+            SubjectAlternativeNameBuilder builder = new SubjectAlternativeNameBuilder();
+
+            // [nihongo]@[nihongo].example.org
+            Assert.Throws<CryptographicException>(
+                () => builder.AddEmailAddress("\u65E5\u672C\u8A8E@\u65E5\u672C\u8A8E.example.org"));
+        }
+
+        [Fact]
+        public static void SingleValue_IPAddress_v4()
+        {
+            SubjectAlternativeNameBuilder builder = new SubjectAlternativeNameBuilder();
+
+            builder.AddIpAddress(IPAddress.Loopback);
+
+            X509Extension extension = builder.Build();
+            Assert.Equal(SubjectAltNameOid, extension.Oid.Value);
+
+            Assert.Equal("300687047F000001", extension.RawData.ByteArrayToHex());
+        }
+
+        [Fact]
+        public static void SingleValue_IPAddress_v6()
+        {
+            SubjectAlternativeNameBuilder builder = new SubjectAlternativeNameBuilder();
+
+            builder.AddIpAddress(IPAddress.IPv6Loopback);
+
+            X509Extension extension = builder.Build();
+            Assert.Equal(SubjectAltNameOid, extension.Oid.Value);
+
+            Assert.Equal("3012871000000000000000000000000000000001", extension.RawData.ByteArrayToHex());
+        }
+
+        [Fact]
+        public static void SingleValue_Uri_Ascii()
+        {
+            SubjectAlternativeNameBuilder builder = new SubjectAlternativeNameBuilder();
+
+            builder.AddUri(new Uri("http://www.example.org/"));
+
+            X509Extension extension = builder.Build();
+            Assert.Equal(SubjectAltNameOid, extension.Oid.Value);
+
+            Assert.Equal(
+                "30198617687474703A2F2F7777772E6578616D706C652E6F72672F",
+                extension.RawData.ByteArrayToHex());
+        }
+
+        [Fact]
+        public static void SingleValue_Uri_UnicodeHost()
+        {
+            SubjectAlternativeNameBuilder builder = new SubjectAlternativeNameBuilder();
```

Comment:
What is it with unicode email characters that makes us want to have a negative test them? I assume the validation is done by the provider\algorithm internal implementation, not higher level?